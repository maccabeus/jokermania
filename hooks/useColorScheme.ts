import { ColorSchemeName, useColorScheme as _useColorScheme } from 'react-native';
import Layout from '../constants/Layout';
import { light, dark } from '../constants/Colors';

export type AppColorScheme = NonNullable<ColorSchemeName>;
export type AppColors = typeof light | typeof dark;
export type Layout = typeof Layout;

export type AppPropKeys = "theme" | any;

export type AppTheme = {
  color: AppColors;
  layout: Layout;
}

export type AppColorSchemeData = {
  [propName: string]: AppColors;
}

export type AppThemeContext = {
  [propName in AppPropKeys]: AppTheme
}
// export type AppThemeContext = {
//   theme: AppTheme
// }

export function useThemeSettings(colorScheme: AppColorScheme): AppTheme {
  return (
    {
      color: colorScheme === "dark" ? dark : light,
      layout: Layout
    }
  );
}

export default function useColorScheme(): AppColorScheme {
  return _useColorScheme() as AppColorScheme;
}