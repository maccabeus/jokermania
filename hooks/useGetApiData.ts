
import { callApi } from '../services/ApiService';
import { useState, useEffect } from 'react';
import { ApiCallStateEnum, IApiGetData } from '../services/types';

export type Predicate = ((...args: any) => boolean) | null;

export const useGetApiData = (apiPath: string, dataIndex: string | null = null, canFetch: Predicate = null): IApiGetData => {
    return fetchApiData(apiPath, null, dataIndex, canFetch);
}

export const fetchApiData = (apiPath: string, queryParam: string | null = null, dataIndex: string | null = null, canFetch: Predicate = null): IApiGetData => {

    const [status, setStatus] = useState(ApiCallStateEnum.IDLE);
    const [data, setData] = useState(undefined);
    const [success, setSuccess] = useState(false);

    useEffect(() => {
        if ((canFetch && canFetch() === true) || !canFetch) {
            /** set loading status */
            const apiFullPath = queryParam ? `${apiPath}=${queryParam}` : apiPath;
            setStatus(ApiCallStateEnum.LOADING);
            callApi(apiFullPath, "GET").then(response => {
                const { success, data } = response;
                if (success) {
                    // if there is a data index to retrieve, dig a bit deeper in  
                    dataIndex ? setData(data[dataIndex]) : setData(data);
                } else {
                    setData(undefined);
                }
                /** set the call success and status values */
                setSuccess(success);
                setStatus(ApiCallStateEnum.COMPLETED);
            }).catch((e: any) => {
                setStatus(ApiCallStateEnum.ERROR);
            })
        }
    }, [apiPath]);

    return { success, status, data, }
}