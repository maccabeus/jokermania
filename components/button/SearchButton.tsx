import { FC } from "react";
import { CardProps } from "../types";
import styled from "styled-components/native";
import { AppThemeContext } from "../../hooks/useColorScheme";
import { BaseOcticons } from "../icon/Octicons";

export const SearchIconView = styled.TouchableOpacity`
    border-radius:${(props: AppThemeContext) => props.theme.layout.fullBorderRadius};
    margin-right:${(props: AppThemeContext) => props.theme.layout.cardMargin};
    align-items:center;
    justify-content:center;
    height:50px;
    width:50px;
    padding: ${(props: AppThemeContext) => props.theme.layout.extraSmallPadding};;
`
const Icon: any = styled(BaseOcticons)`
    opacity:0.9;
`;

const SearchButton: FC<CardProps> = (props) => {
    return (
        <SearchIconView  {...props}>
            <Icon name="search" size={24} />
        </SearchIconView >
    )
}
export default SearchButton;