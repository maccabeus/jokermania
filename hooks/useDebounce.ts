import { Predicate } from "./useGetApiData";

const debounce = (waitTime: number = 1000, maxKeyStrokes: number = 3) => {

    let timerInterval: any = null;

    const debounceAction = (callback: Function, actionPredicate: Predicate) => {
        if (timerInterval) clearTimeout(timerInterval);

        if (actionPredicate && actionPredicate() === true) {
            timerInterval = setTimeout(() => {
                callback();
            }, waitTime);
        }
    }
    return debounceAction
}
const debounceOld = (waitTime: number = 2000, maxKeyStrokes: number = 3) => {

    let keyCount: number = 0;
    let timerInterval: any = null;

    const debounceAction = (callback: Function, actionPredicate: boolean = true) => {
        if (timerInterval) clearTimeout(timerInterval);
        keyCount += 1;
        console.log("key count", keyCount);
        if (actionPredicate) {
            timerInterval = setTimeout(() => {
                callback();
                /** reset keyCount */
                keyCount = 0;
                console.log("calling now count: ", keyCount);
            }, waitTime);
        }
    }
    return debounceAction
}

export default debounce;