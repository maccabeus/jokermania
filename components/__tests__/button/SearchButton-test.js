import * as React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components/native';
import { useThemeSettings } from '../../../hooks/useColorScheme';
import { jest } from '@jest/globals';
import SearchButton from '../../button/SearchButton';

describe("Search Button Snapshot", () => {

  const scheme = jest.fn();
  let lightScheme = scheme.mockReturnValue("light");
  let currentScheme = null;

  beforeEach(() => {
    currentScheme = useThemeSettings(lightScheme)
  });

  test(`Search button renders correctly`, () => {
    const tree = renderer.create(
      <ThemeProvider theme={currentScheme}>
        <SearchButton />
      </ThemeProvider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});


