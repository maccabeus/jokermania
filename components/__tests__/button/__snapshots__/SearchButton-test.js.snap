// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Button Snapshot Search button renders correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "borderRadius": 3000,
      "height": 50,
      "justifyContent": "center",
      "marginRight": 10,
      "opacity": 1,
      "paddingBottom": 10,
      "paddingLeft": 10,
      "paddingRight": 10,
      "paddingTop": 10,
      "width": 50,
    }
  }
>
  <Text />
</View>
`;
